Agent forwarding implementation

MichaelBuckley opened the pull request:
This PR contains a series of patches that date back many years and I believe were discussed on the mailing list, but never merged. We have been using these in our local copy of libssh2 without issue since 2015, if not earlier. I believe this is the full set of changes, as we tried to use comments to mark where our copy of libssh2 differs from the canonical version.

This also contains changes I made earlier this year, but which were not discussed on the mailing list, to support certificates and FIDO2 keys with agent forwarding.

Note that this is not a complete implementation of agent forwarding, as that is outside the scope of libssh2. Clients still need to provide their own implementation that parses ssh-agent methods after calling libssh2_channel_read() and calls the appropriate callback messages in libssh2. See the man page changes in this PR for more details.

MichaelBuckley committed:
Copy agent authentication code from mailing list patches

MichaelBuckley committed:
Add man page for libssh2_agent_userauth(3)

MichaelBuckley committed:
Add support for certificate signing with ssh-agent

MichaelBuckley committed:
Move declaration to top of block

MichaelBuckley committed:
Move declarations to top of block

vszakats committed:
Merge branch 'master' into agent-forwarding-implementation

vszakats commented on src/packet.c:
```diff
@@ -460,6 +460,156 @@ packet_x11_open(LIBSSH2_SESSION * session, unsigned char *data,
+
+            authagent_state->state = libssh2_NB_state_created;
+        }
+
+        if(authagent_state->state == libssh2_NB_state_created) {
+            rc = _libssh2_transport_send(session, authagent_state->packet, 17,
+                         NULL, 0);
+            if(rc == LIBSSH2_ERROR_EAGAIN) {
+                return rc;
+            }
+            else if(rc) {
+                authagent_state->state = libssh2_NB_state_idle;
+                return _libssh2_error(session, LIBSSH2_ERROR_SOCKET_SEND,
+                            "Unable to send channel open "
+                            "confirmation");
+            }
+
+            /* Link the channel into the session */
+            _libssh2_list_add(&session->channels, &channel->node);
+
+            /* mess with stuff so we don't keep reading the same packet
+               over and over */
+            session->packet.total_num = 0;
+            session->fullpacket_state = libssh2_NB_state_idle;
+
+            /*
+            * Pass control to the callback, they may turn right around and
+            * free the channel, or actually use it
+            */
+
+            LIBSSH2_AUTHAGENT(channel);
+
+            authagent_state->state = libssh2_NB_state_idle;
+            return 0;
+        }
+  }
+    else
+        failure_code = SSH_OPEN_RESOURCE_SHORTAGE;
+
+    /* fall-trough */
```
```suggestion
    /* fall-through */
```

vszakats replied:
```suggestion
    if(session->authagent) {
        if(authagent_state->state == libssh2_NB_state_allocated) {
            channel = LIBSSH2_ALLOC(session, sizeof(LIBSSH2_CHANNEL));
            authagent_state->channel = channel;

            if(!channel) {
                _libssh2_error(session, LIBSSH2_ERROR_ALLOC,
                               "allocate a channel for new connection");
                failure_code = SSH_OPEN_RESOURCE_SHORTAGE;
                goto authagent_exit;
            }
            memset(channel, 0, sizeof(LIBSSH2_CHANNEL));

            channel->session = session;
            channel->channel_type_len = sizeof("auth agent") - 1;
            channel->channel_type = LIBSSH2_ALLOC(session,
                                                  channel->channel_type_len +
                                                  1);
            if(!channel->channel_type) {
                _libssh2_error(session, LIBSSH2_ERROR_ALLOC,
                               "allocate a channel for new connection");
                LIBSSH2_FREE(session, channel);
                failure_code = SSH_OPEN_RESOURCE_SHORTAGE;
                goto authagent_exit;
            }
            memcpy(channel->channel_type, "auth agent",
                   channel->channel_type_len + 1);

            channel->remote.id = authagent_state->sender_channel;
            channel->remote.window_size_initial =
                LIBSSH2_CHANNEL_WINDOW_DEFAULT;
            channel->remote.window_size = LIBSSH2_CHANNEL_WINDOW_DEFAULT;
            channel->remote.packet_size = LIBSSH2_CHANNEL_PACKET_DEFAULT;

            channel->local.id = _libssh2_channel_nextid(session);
            channel->local.window_size_initial =
                authagent_state->initial_window_size;
            channel->local.window_size = authagent_state->initial_window_size;
            channel->local.packet_size = authagent_state->packet_size;

            _libssh2_debug(session, LIBSSH2_TRACE_CONN,
                           "Auth Agent Connection established: channel "
                           "%lu/%lu win %lu/%lu packet %lu/%lu",
                           channel->local.id, channel->remote.id,
                           channel->local.window_size,
                           channel->remote.window_size,
                           channel->local.packet_size,
                           channel->remote.packet_size);

            p = authagent_state->packet;
            *(p++) = SSH_MSG_CHANNEL_OPEN_CONFIRMATION;
            _libssh2_store_u32(&p, channel->remote.id);
            _libssh2_store_u32(&p, channel->local.id);
            _libssh2_store_u32(&p, channel->remote.window_size_initial);
            _libssh2_store_u32(&p, channel->remote.packet_size);

            authagent_state->state = libssh2_NB_state_created;
        }

        if(authagent_state->state == libssh2_NB_state_created) {
            rc = _libssh2_transport_send(session, authagent_state->packet, 17,
                                         NULL, 0);
            if(rc == LIBSSH2_ERROR_EAGAIN) {
                return rc;
            }
            else if(rc) {
                authagent_state->state = libssh2_NB_state_idle;
                return _libssh2_error(session, LIBSSH2_ERROR_SOCKET_SEND,
                                      "Unable to send channel open "
                                      "confirmation");
            }

            /* Link the channel into the session */
            _libssh2_list_add(&session->channels, &channel->node);

            /* mess with stuff so we don't keep reading the same packet
               over and over */
            session->packet.total_num = 0;
            session->fullpacket_state = libssh2_NB_state_idle;

            /* Pass control to the callback, they may turn right around and
               and free the channel, or actually use it */

            LIBSSH2_AUTHAGENT(channel);

            authagent_state->state = libssh2_NB_state_idle;
            return 0;
        }
    }
    else
        failure_code = SSH_OPEN_RESOURCE_SHORTAGE;

    /* fall-through */
```

vszakats commented on src/userauth.c:
```diff
@@ -1701,8 +1724,8 @@ _libssh2_userauth_publickey(LIBSSH2_SESSION *session,
         session->userauth_pblc_method_len =
-           plain_method_len((const char *)session->userauth_pblc_method,
-                            session->userauth_pblc_method_len);
+        plain_method((char *)session->userauth_pblc_method,
+                     session->userauth_pblc_method_len);
```
```suggestion
            plain_method((char *)session->userauth_pblc_method,
                         session->userauth_pblc_method_len);
```

vszakats commented on docs/libssh2_agent_sign.3:
```diff
@@ -0,0 +1,50 @@
+.SH NAME
+libssh2_agent_sign - sign data, with the help of ssh-agent
+.SH SYNOPSIS
+#include <libssh2.h>
+
+int libssh2_agent_sign(LIBSSH2_AGENT *agent,
+                       struct libssh2_agent_publickey *identity,
+                       unsigned char **sig,
+                       size_t *s_len,
+                       unsigned char *data,
+                       size_t d_len,
+                       const char *method,
+                       u_int method_len);
+
```
```suggestion
.nf
#include <libssh2.h>

int
libssh2_agent_sign(LIBSSH2_AGENT *agent,
                   struct libssh2_agent_publickey *identity,
                   unsigned char **sig,
                   size_t *s_len,
                   unsigned char *data,
                   size_t d_len,
                   const char *method,
                   u_int method_len);
.fi
```

vszakats commented on docs/libssh2_agent_sign.3:
```diff
@@ -0,0 +1,50 @@
+.SH RETURN VALUE
+Returns 0 if succeeded, or a negative value for error.
+.SH AVAILABILITY
+Added in libssh2 1.2
```
```suggestion
Added in libssh2 1.11.0
```

vszakats commented on docs/libssh2_agent_sign.3:
```diff
@@ -0,0 +1,50 @@
+.TH libssh2_agent_sign 3 "1 Oct 2022" "libssh2 1.10.0" "libssh2 manual"
```
```suggestion
.TH libssh2_agent_sign 3 "1 Oct 2022" "libssh2 1.11.0" "libssh2 manual"
```

vszakats committed:
docs

vszakats commented on docs/libssh2_session_callback_set.3:
```diff
@@ -71,8 +71,62 @@ to the abstract pointer set in the \fIlibssh2_session_init_ex(3)\fP call.
+The prototype of the callback:
+
+.nf
+void identities(LIBSSH2_SESSION* session, void *buffer, const char *agent_path,
+                void **abstract)
```
```suggestion
void identities(LIBSSH2_SESSION* session, void *buffer,
                const char *agent_path,
                void **abstract)
```

vszakats commented on docs/libssh2_session_callback_set.3:
```diff
@@ -71,8 +71,63 @@ to the abstract pointer set in the \fIlibssh2_session_init_ex(3)\fP call.
+The prototype of the callback:
+
+.nf
+void sign(LIBSSH2_SESSION *session, unsigned char **sig, size_t *sig_len,
+   const unsigned char *data, size_t data_len, void **abstract);
```
```suggestion
void sign(LIBSSH2_SESSION *session,
          unsigned char **sig, size_t *sig_len,
          const unsigned char *data, size_t data_len,
          void **abstract);
```

vszakats commented on src/agent.c:
```diff
@@ -478,8 +479,12 @@ agent_sign(LIBSSH2_SESSION *session, unsigned char **sig, size_t *sig_len,
+
     /* check to see if we match requested */
-    if((size_t)method_len != session->userauth_pblc_method_len ||
+    if(((size_t)method_len != session->userauth_pblc_method_len &&
+        method_len != plain_len) ||
```
```suggestion
    if(((size_t)method_len != session->userauth_pblc_method_len &&
       method_len != plain_len) ||
```

vszakats committed:
indent

vszakats commented on include/libssh2.h:
```diff
@@ -610,6 +626,7 @@ LIBSSH2_API int libssh2_banner_set(LIBSSH2_SESSION *session,
 LIBSSH2_API int libssh2_session_startup(LIBSSH2_SESSION *session, int sock);
 LIBSSH2_API int libssh2_session_handshake(LIBSSH2_SESSION *session,
                                           libssh2_socket_t sock);
+
```
```suggestion
```

vszakats commented on src/agent.c:
```diff
@@ -478,8 +479,12 @@ agent_sign(LIBSSH2_SESSION *session, unsigned char **sig, size_t *sig_len,
+
     /* check to see if we match requested */
-    if((size_t)method_len != session->userauth_pblc_method_len ||
+    if(((size_t)method_len != session->userauth_pblc_method_len &&
+       method_len != plain_len) ||
         memcmp(method_name, session->userauth_pblc_method, method_len)) {
```
```suggestion
    if(((size_t)method_len != session->userauth_pblc_method_len &&
        method_len != plain_len) ||
       memcmp(method_name, session->userauth_pblc_method, method_len)) {
```

vszakats commented on src/libssh2_priv.h:
```diff
@@ -196,6 +196,18 @@ struct iovec {
     channel->session->x11(((channel)->session), (channel), \
                           (shost), (sport), (&(channel)->session->abstract))
 
+#define LIBSSH2_AUTHAGENT(channel) \
+    channel->session->authagent(((channel)->session), (channel), \
+    (&(channel)->session->abstract))
+
+#define LIBSSH2_ADD_IDENTITIES(session, buffer, agentPath) \
+    session->addLocalIdentities((session), (buffer), (agentPath), \
+    (&(session->abstract)))
+
+#define LIBSSH2_AUTHAGENT_SIGN(session, blob, blen, data, dlen, sig, sigLen, \
+    agentPath) session->agentSignCallback((session), (blob), (blen), \
+    (data), (dlen), (sig), (sigLen), (agentPath), (&(session->abstract)))
+
```
```suggestion
#define LIBSSH2_AUTHAGENT(channel) \
    channel->session->authagent(((channel)->session), (channel), \
                                (&(channel)->session->abstract))

#define LIBSSH2_ADD_IDENTITIES(session, buffer, agentPath) \
    session->addLocalIdentities((session), (buffer), \
                                (agentPath), (&(session->abstract)))

#define LIBSSH2_AUTHAGENT_SIGN(session, blob, blen, \
                               data, dlen, sig, sigLen, \
                               agentPath) \
    session->agentSignCallback((session), (blob), (blen), \
                               (data), (dlen), (sig), (sigLen), \
                               (agentPath), (&(session->abstract)))

```

vszakats commented on src/libssh2_priv.h:
```diff
@@ -341,6 +357,25 @@ typedef struct packet_x11_open_state_t
+
+typedef enum
+{
+  libssh2_requires_size_decryption = (1 << 0),
+  libssh2_requires_size_field_in_packet = (1 << 1)
```
```suggestion
    libssh2_requires_size_decryption = (1 << 0),
    libssh2_requires_size_field_in_packet = (1 << 1)
```

vszakats commented on src/packet.c:
```diff
@@ -1054,6 +1204,20 @@ _libssh2_packet_add(LIBSSH2_SESSION * session, unsigned char *data,
                 rc = packet_x11_open(session, data, datalen,
                                      &session->packAdd_x11open_state);
             }
+            else if((datalen >= (sizeof("auth-agent@openssh.com") + 4)) &&
+                     ((sizeof("auth-agent@openssh.com") - 1) ==
+                        _libssh2_ntohu32(data + 1)) &&
+                     (memcmp(data + 5, "auth-agent@openssh.com",
+                        sizeof("auth-agent@openssh.com") - 1) == 0)) {
+
+                /* init the state struct */
+                memset(&session->packAdd_authagent_state, 0,
+                       sizeof(session->packAdd_authagent_state));
+
+                libssh2_packet_add_jump_authagent:
+                    rc = packet_authagent_open(session, data, datalen,
+                        &session->packAdd_authagent_state);
+            }
```
```suggestion
            else if((datalen >= (sizeof("auth-agent@openssh.com") + 4)) &&
                    ((sizeof("auth-agent@openssh.com") - 1) ==
                      _libssh2_ntohu32(data + 1)) &&
                    (memcmp(data + 5, "auth-agent@openssh.com",
                            sizeof("auth-agent@openssh.com") - 1) == 0)) {

                /* init the state struct */
                memset(&session->packAdd_authagent_state, 0,
                       sizeof(session->packAdd_authagent_state));

libssh2_packet_add_jump_authagent:

                rc = packet_authagent_open(session, data, datalen,
                                           &session->packAdd_authagent_state);
            }
```

vszakats commented on src/agent.c:
```diff
@@ -829,6 +834,57 @@ libssh2_agent_userauth(LIBSSH2_AGENT *agent,
+    uint32_t methodLen;
+
+    if(agent->session->userauth_pblc_state == libssh2_NB_state_idle) {
+        memset(&agent->transctx, 0, sizeof agent->transctx);
```
```suggestion
        memset(&agent->transctx, 0, sizeof(agent->transctx));
```

vszakats committed:
checksrc

vszakats commented on src/userauth.c:
```diff
@@ -2291,56 +2314,110 @@ LIBSSH2_API int
                                   "Unable to extract public key "
                                   "from private key.");
+        }
+        else if(publickeydata_len == 0 || publickeydata == NULL) {
+            session->userauth_pblc_method = tmp_method;
+            session->userauth_pblc_method_len = tmp_method_len;
+
+            pubkeydata_len = tmp_publickeydata_len;
+            pubkeydata = tmp_publickeydata;
+        }
+        else {
+            const char *ecdsa = "sk-ecdsa-sha2-nistp256-cert-v01@openssh.com";
+            const char *ed25519 = "sk-ssh-ed25519-cert-v01@openssh.com";
+
+            if(tmp_method != NULL) {
```
```suggestion
        else if(publickeydata_len == 0 || !publickeydata) {
            session->userauth_pblc_method = tmp_method;
            session->userauth_pblc_method_len = tmp_method_len;

            pubkeydata_len = tmp_publickeydata_len;
            pubkeydata = tmp_publickeydata;
        }
        else {
            const char *ecdsa = "sk-ecdsa-sha2-nistp256-cert-v01@openssh.com";
            const char *ed25519 = "sk-ssh-ed25519-cert-v01@openssh.com";

            if(tmp_method) {
```

vszakats commented on src/packet.c:
```diff
@@ -460,6 +460,154 @@ packet_x11_open(LIBSSH2_SESSION * session, unsigned char *data,
+                failure_code = SSH_OPEN_RESOURCE_SHORTAGE;
+                goto authagent_exit;
+            }
+            memset(channel, 0, sizeof(LIBSSH2_CHANNEL));
+
+            channel->session = session;
+            channel->channel_type_len = sizeof("auth agent") - 1;
+            channel->channel_type = LIBSSH2_ALLOC(session,
+                                                  channel->channel_type_len +
+                                                  1);
+            if(!channel->channel_type) {
+                _libssh2_error(session, LIBSSH2_ERROR_ALLOC,
+                               "allocate a channel for new connection");
+                LIBSSH2_FREE(session, channel);
+                failure_code = SSH_OPEN_RESOURCE_SHORTAGE;
+                goto authagent_exit;
+            }
+            memcpy(channel->channel_type, "auth agent",
+                   channel->channel_type_len + 1);
+
+            channel->remote.id = authagent_state->sender_channel;
+            channel->remote.window_size_initial =
+                LIBSSH2_CHANNEL_WINDOW_DEFAULT;
+            channel->remote.window_size = LIBSSH2_CHANNEL_WINDOW_DEFAULT;
+            channel->remote.packet_size = LIBSSH2_CHANNEL_PACKET_DEFAULT;
+
+            channel->local.id = _libssh2_channel_nextid(session);
+            channel->local.window_size_initial =
+                authagent_state->initial_window_size;
+            channel->local.window_size = authagent_state->initial_window_size;
+            channel->local.packet_size = authagent_state->packet_size;
+
+            _libssh2_debug(session, LIBSSH2_TRACE_CONN,
+                           "Auth Agent Connection established: channel "
+                           "%lu/%lu win %lu/%lu packet %lu/%lu",
+                           channel->local.id, channel->remote.id,
+                           channel->local.window_size,
+                           channel->remote.window_size,
+                           channel->local.packet_size,
+                           channel->remote.packet_size);
```
```suggestion
        _libssh2_debug((session, LIBSSH2_TRACE_CONN,
                       "Auth Agent Connection Received on channel %lu",
                       authagent_state->sender_channel));

        authagent_state->state = libssh2_NB_state_allocated;
    }

    if(session->authagent) {
        if(authagent_state->state == libssh2_NB_state_allocated) {
            channel = LIBSSH2_ALLOC(session, sizeof(LIBSSH2_CHANNEL));
            authagent_state->channel = channel;

            if(!channel) {
                _libssh2_error(session, LIBSSH2_ERROR_ALLOC,
                               "allocate a channel for new connection");
                failure_code = SSH_OPEN_RESOURCE_SHORTAGE;
                goto authagent_exit;
            }
            memset(channel, 0, sizeof(LIBSSH2_CHANNEL));

            channel->session = session;
            channel->channel_type_len = sizeof("auth agent") - 1;
            channel->channel_type = LIBSSH2_ALLOC(session,
                                                  channel->channel_type_len +
                                                  1);
            if(!channel->channel_type) {
                _libssh2_error(session, LIBSSH2_ERROR_ALLOC,
                               "allocate a channel for new connection");
                LIBSSH2_FREE(session, channel);
                failure_code = SSH_OPEN_RESOURCE_SHORTAGE;
                goto authagent_exit;
            }
            memcpy(channel->channel_type, "auth agent",
                   channel->channel_type_len + 1);

            channel->remote.id = authagent_state->sender_channel;
            channel->remote.window_size_initial =
                LIBSSH2_CHANNEL_WINDOW_DEFAULT;
            channel->remote.window_size = LIBSSH2_CHANNEL_WINDOW_DEFAULT;
            channel->remote.packet_size = LIBSSH2_CHANNEL_PACKET_DEFAULT;

            channel->local.id = _libssh2_channel_nextid(session);
            channel->local.window_size_initial =
                authagent_state->initial_window_size;
            channel->local.window_size = authagent_state->initial_window_size;
            channel->local.packet_size = authagent_state->packet_size;

            _libssh2_debug((session, LIBSSH2_TRACE_CONN,
                           "Auth Agent Connection established: channel "
                           "%lu/%lu win %lu/%lu packet %lu/%lu",
                           channel->local.id, channel->remote.id,
                           channel->local.window_size,
                           channel->remote.window_size,
                           channel->local.packet_size,
                           channel->remote.packet_size));
```

vszakats committed:
update debug calls

vszakats commented on docs/Makefile.am:
```diff
@@ -10,6 +10,7 @@ dist_man_MANS = \
   libssh2_agent_init.3 \
   libssh2_agent_list_identities.3 \
   libssh2_agent_set_identity_path.3 \
+  libssh2_agent_sign.3
```
```suggestion
  libssh2_agent_sign.3 \
```

vszakats commented:
Turns we have a PR for `libssh2_agent_sign()` with a different implementation that is just a wrapper to the internal function: #516

MichaelBuckley commented:
I have updated the title to remove the `WIP`. It was no longer accurate. Thank you for catching that.

I would advocate that this should be preferred over #516. I opened this PR after @willco007 and I chatted about getting some of our changes merged back upstream, and he mentioned that #516 was not as complete.

vszakats committed:
fix compiler warning

src\packet.c(1215,59): warning C4267: 'function': conversion from 'size_t' to 'unsigned long', possible loss of data

vszakats commented on src/packet.c:
```diff
@@ -460,6 +460,154 @@ packet_x11_open(LIBSSH2_SESSION * session, unsigned char *data,
+ */
+static inline int
+packet_authagent_open(LIBSSH2_SESSION * session, unsigned char *data,
+                      unsigned long datalen,
```
```suggestion
                      size_t datalen,
```

vszakats committed:
prefer size_t

vszakats commented on include/libssh2.h:
```diff
@@ -1324,6 +1345,23 @@ libssh2_agent_userauth(LIBSSH2_AGENT *agent,
+                   unsigned char *data,
+                   size_t d_len,
+                   const char *method,
+                   u_int method_len);
```
```suggestion
                   unsigned int method_len);
```

vszakats committed:
prefer unsigned int over u_int in public API

vszakats commented on docs/libssh2_agent_sign.3:
```diff
@@ -0,0 +1,52 @@
+                   unsigned char *data,
+                   size_t d_len,
+                   const char *method,
+                   u_int method_len);
```
```suggestion
                   unsigned int method_len);
```

vszakats commented on src/agent.c:
```diff
@@ -829,6 +834,57 @@ libssh2_agent_userauth(LIBSSH2_AGENT *agent,
+                   unsigned char *data,
+                   size_t d_len,
+                   const char *method,
+                   u_int method_len)
```
```suggestion
                   unsigned int method_len)
```

vszakats commented on docs/libssh2_agent_sign.3:
```diff
@@ -0,0 +1,52 @@
+                   struct libssh2_agent_publickey *identity,
+                   unsigned char **sig,
+                   size_t *s_len,
+                   unsigned char *data,
```
```suggestion
                   const unsigned char *data,
```

vszakats committed:
add const

vszakats commented on include/libssh2.h:
```diff
@@ -1324,6 +1345,23 @@ libssh2_agent_userauth(LIBSSH2_AGENT *agent,
+                   struct libssh2_agent_publickey *identity,
+                   unsigned char **sig,
+                   size_t *s_len,
+                   unsigned char *data,
```
```suggestion
                   const unsigned char *data,
```

vszakats commented on src/agent.c:
```diff
@@ -829,6 +834,57 @@ libssh2_agent_userauth(LIBSSH2_AGENT *agent,
+                   struct libssh2_agent_publickey *identity,
+                   unsigned char **sig,
+                   size_t *s_len,
+                   unsigned char *data,
```
```suggestion
                   const unsigned char *data,
```

vszakats commented on src/packet.c:
```diff
@@ -460,6 +460,154 @@ packet_x11_open(LIBSSH2_SESSION * session, unsigned char *data,
+ * Open a connection to authentication agent
+ */
+static inline int
+packet_authagent_open(LIBSSH2_SESSION * session, unsigned char *data,
+                      size_t datalen,
```
```suggestion
packet_authagent_open(LIBSSH2_SESSION * session,
                      const unsigned char *data, size_t datalen,
```

vszakats committed:
fmt

vszakats commented on include/libssh2.h:
```diff
@@ -333,6 +333,22 @@ typedef struct _LIBSSH2_SK_SIG_INFO {
+#define LIBSSH2_AUTHAGENT_SIGN_FUNC(name) \
+   int name(LIBSSH2_SESSION* session, \
+            unsigned char *blob, unsigned int blen, \
+            unsigned char *data, unsigned int dlen, \
```
```suggestion
            const unsigned char *data, unsigned int dlen, \
```

vszakats committed:
add const (hopefully OK)

vszakats commented:
@MichaelBuckley Thank you for your update!

We're mostly ready to merge, but it seems the documented callback arguments for `LIBSSH2_CALLBACK_AUTHAGENT_IDENTITIES` and `LIBSSH2_CALLBACK_AUTHAGENT_SIGN` differ from the ones in the code, or am I misreading? (from sign, blob seems to be missing.)

vszakats commented on include/libssh2.h:
```diff
@@ -333,6 +333,22 @@ typedef struct _LIBSSH2_SK_SIG_INFO {
+#define LIBSSH2_AUTHAGENT_SIGN_FUNC(name) \
+   int name(LIBSSH2_SESSION* session, \
+            unsigned char *blob, unsigned int blen, \
+            const unsigned char *data, unsigned int dlen, \
```
@MichaelBuckley I've added `const` here. Do you think this is okay?

MichaelBuckley replied:
Yes, good catch. However, we will also need to add const to `libssh2_agent_sign()` to match. The `agent_sign()` function already marks the data parameter `const`.

vszakats replied:
Yep, added that one already. Though we still need to sync the man page with this.

vszakats commented on docs/libssh2_session_callback_set.3:
```diff
@@ -71,8 +71,65 @@ to the abstract pointer set in the \fIlibssh2_session_init_ex(3)\fP call.
+void sign(LIBSSH2_SESSION *session,
+          unsigned char **sig, size_t *sig_len,
+          const unsigned char *data, size_t data_len,
+          void **abstract);
```
The callback arguments here (`LIBSSH2_CALLBACK_AUTHAGENT_IDENTITIES` and `LIBSSH2_CALLBACK_AUTHAGENT_SIGN`) seem to differ from the ones in the code, or am I misreading?

For example from the SIGN callback, blob seems to be missing and `sign` comes before `data` here, but after in the code.

MichaelBuckley replied:
You're right about `LIBSSH2_CALLBACK_AUTHAGENT_SIGN`. I think it was changed after I last edited the man page.

It should be `int sign(LIBSSH2_SESSION* session, unsigned int flags, unsigned char *blob, unsigned int blen, unsigned char *data, const unsigned int dlen, unsigned char **sig, unsigned int *sig_len, const char *agent_path, void **abstract);`

I'll update the man page in a bit here.

`LIBSSH2_CALLBACK_AUTHAGENT_IDENTITIES` looks correct to me though, unless I'm missing something.

vszakats replied:
`sign()` will be fine then, thanks for looking into it.

As for `identities`, I'm looking at this code in `include/libssh2.h`:
```
#define LIBSSH2_ADD_IDENTITIES_FUNC(name) \
   void name(LIBSSH2_SESSION *session, LIBSSH2_CHANNEL *channel, \
             void **abstract)
```
This doesn't match `LIBSSH2_CALLBACK_AUTHAGENT_IDENTITIES` in 
the manual or `LIBSSH2_ADD_IDENTITIES()` from `libssh2_priv.h`. They 
look like the same callback, though I may be wrong.

MichaelBuckley replied:
Oops, I was looking at the wrong branch. The identities in the man file was correct, but it was incorrect in the header. I was looking at a branch where it was correct in the header.

Both problems should now be fixed. The man file was updated with the correct `sign()`, and the header was updated to match `identities()` in the man file.

vszakats replied:
Thanks, no worries. We're good to merge once the CI tests finished.

vszakats committed:
indent in libssh2.h

MichaelBuckley committed:
Make the docs and the implementation match

MichaelBuckley committed:
More indentation

vszakats committed:
doc fixup const

vszakats commented:
Merged now, thank you @MichaelBuckley!

vszakats commented on src/userauth.c:
```diff
@@ -1268,6 +1268,29 @@ static size_t plain_method_len(const char *method, size_t method_len)
+                method,
+                method_len)) {
+        const char *new_method = "sk-ecdsa-sha2-nistp256@openssh.com";
+        memcpy(method, new_method, strlen(new_method));
```
@MichaelBuckley clang-tidy and Copilot draw attention to the line above,
and the similar one in the `if` following. The new string is copied over the
input one, but without a zero-terminator. Is it intended that these two inputs
get overwritten, while the others don't? If so, wouldn't we need to copy the
NUL byte too?

(It also confused me why these two values overwrite the input, and return the
full length, while pre-existing ones return the length of the prefix without
touching the buffer.)

https://github.com/libssh2/libssh2/security/quality/ai-findings (may disappear from this page):
"The memcpy operation copies exactly strlen(new_method) bytes without null-termination. The comment suggests this is intentional, but the function returns strlen(new_method) which doesn't account for a null terminator. If the caller expects a null-terminated string (which the 'char *method' parameter suggests), this could lead to issues. Either ensure null-termination or document that the returned buffer is not null-terminated."

clang-tidy output can be seen by dropping the NOLINTLINT comments, I did
not save them unfortunately.

vszakats replied:
I double-checked and the code is fine. That's assuming all callers honor the returned length, and never consider the buffer as nul-terminated. Which was already an assumption made before this patch.

Thus, sorry about the noise and disregard the message above.
